File: 力扣/学习计划/剑指 Offer I/069_丑数/main.go
package main

func nthUglyNumber(n int) int {
	var dp []int = make([]int, n)
	dp[0] = 1 // 最基本的丑数
	i, j, k := 0, 0, 0
	for idx := 1; idx < n; idx++ {
		tmp := minNum(dp[i]*2, dp[j]*3, dp[k]*5)
		if tmp == dp[i]*2 {
			i++
		}
		if tmp == dp[j]*3 {
			j++
		}
		if tmp == dp[k]*5 {
			k++
		}
		dp[idx] = tmp
	}
	return dp[n-1]
}

// 获取 num1, num2, num3 中的最小值
func minNum(num1, num2, num3 int) int {
	if num2 < num1 {
		num1 = num2
	}
	if num3 < num1 {
		return num3
	}
	return num1
}

File: 力扣/学习计划/剑指 Offer I/069_丑数/main.java
import javax.management.openmbean.SimpleType;

/**
 * https://leetcode-cn.com/problems/chou-shu-lcof/
 * 动态规划
 */
class Solution {
    public int nthUglyNumber(int n) {
        int dp[] = new int[n];
        dp[0] = 1; // 基础丑数
        int i = 0, j = 0, k = 0;
        for (int idx = 1; idx < n; idx++) {
            // 任意一个丑数都是由小于它的某一个丑数 *2, *3 或者 *5 得到的
            int tmp = Math.min(dp[i] * 2, Math.min(dp[j] * 3, dp[k] * 5));
            if (tmp == dp[i] * 2) i++;
            if (tmp == dp[j] * 3) j++;
            if (tmp == dp[k] * 5) k++;
            dp[idx] = tmp;
        }
        return dp[n - 1];
    }
}